File: brute_force/java/leetcode/Permutation/Permutation.java
import java.util.*;

import java.util.List;
import java.util.ArrayList;

public class Permutation {
    public static void main(String[] args) {
        Solution sol = new Solution();
        // Test cases here
        System.out.println(sol.permute(new int[] { 1, 2, 3 }));
    }
}
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        List<Integer> path = new ArrayList<>();
        backtrack(nums, path, result);
        return result;
    }
    private void backtrack(int[] nums, List<Integer> path, List<List<Integer>> result) {
        if (path.size() == nums.length) {
            result.add(new ArrayList<>(path));
            return;
        }
        for (int num : nums) {
            if (path.contains(num)) {
                continue;
            }
            path.add(num);
            backtrack(nums, path, result);
            path.remove(path.size() - 1);
        }
    }

}

File: brute_force/python/leetcode/permutation.py
from typing import List


class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        def backtracking(path: List[int]):
            print(path)

            if len(path) == len(nums):
                # result.append(deepcopy(path))
                result.append(path[:])
                return

            for num in nums:
                if num not in path:
                    path.append(num)
                    backtracking(path)
                    path.pop()

        result = []
        backtracking(path=[])
        return result


if __name__ == '__main__':
    nums = [1, 2]
    solution = Solution()
    ans = solution.permute(nums=nums)
    print(ans)
